distributed/worker.py

```diff
@@ -3041,15 +3094,16 @@ def update_who_has(self, who_has, *, stimulus_id):
                         # Do not mutate the input dict. That's rude
                         workers = set(workers) - {self.address}
                     dep_ts = self.tasks[dep]
-                    dep_ts.who_has.update(workers)
+                    if dep_ts.state in FETCH_INTENDED:
+                        dep_ts.who_has.update(workers)
 
-                    if dep_ts.state == "missing":
-                        recommendations[dep_ts] = "fetch"
+                        if dep_ts.state == "missing":
+                            recommendations[dep_ts] = "fetch"
 
-                    for worker in workers:
-                        self.has_what[worker].add(dep)
-                        if dep_ts.state in ("fetch", "flight", "missing"):
-                            self.pending_data_per_worker[worker].append(dep_ts.key)
+                        for worker in workers:
+                            self.has_what[worker].add(dep)
+                            if dep_ts.state in FETCH_INTENDED:
```

Comment:
This if clause shouldn't be necessary any longer since it is true for the entire code block

Reply:
indeed - removing it